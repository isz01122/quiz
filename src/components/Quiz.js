import React from "react";
import {
  Button,
  FormControl,
  FormControlLabel,
  Radio,
  RadioGroup,
  FormHelperText,
  TextField
} from "@material-ui/core";
import { makeStyles } from "@material-ui/core/styles";
import { Bar } from "react-chartjs-2";
import moment from "moment";

const Quiz = ({
  quizzes,
  onFinishPress,
  onNextPress,
  index,
  onOptionChange,
  selectedOption,
  isFinish,
  time,
  onRestartPress,
  onOnceMorePress,
  isWrongAnswerMode,
  onWrongAnswerNotePress,
  onWrongAnswerNoteTextChange
}) => {
  const classes = useStyles();
  const quiz = quizzes[index];
  const isLast = index === quizzes.length - 1 ? true : false;
  const chartData = quizzes.map(quiz =>
    quiz.isCorrect && !quiz.isOnceMore
      ? 100
      : quiz.isCorrect && quiz.isOnceMore
      ? 50
      : 0
  );
  const chartDataColor = chartData.map(value =>
    value === 100 ? "#66DE93" : value === 50 ? "#FF616D" : ""
  );

  const options = {
    legend: {
      display: false
    },
    scales: {
      xAxes: [
        {
          ticks: {
            fontSize: 20,
            fontStyle: "bold",
            fontColor: "#000"
          }
        }
      ],
      yAxes: [
        {
          ticks: {
            min: 0,
            max: 100,
            beginAtZero: true,
            stepSize: 50,
            fontSize: 20,
            fontStyle: "bold",
            fontColor: "#000",
            callback: function (value) {
              return `${value}%`;
            }
          }
        }
      ]
    },
    tooltips: {
      callbacks: {
        label: function (tooltipItems, data) {
          const value = data.datasets[0].data[tooltipItems.index].toFixed(0);
          return `${value}%`;
        }
      }
    },
    maintainAspectRatio: false
  };

  const data = {
    labels: quizzes.map((quiz, index) => `${index + 1}번`),
    datasets: [
      {
        label: "정답률",
        maxBarThickness: 30,
        backgroundColor: chartDataColor,
        data: chartData
      }
    ]
  };

  return (
    <div
      className="quiz-container"
      data-testid={
        !isWrongAnswerMode
          ? "quiz-component"
          : "quiz-component-wrong-answer-mode"
      }
    >
      {!isFinish && (
        <div className="quiz-form">
          <div>
            <div className="text-title">{`Q${index + 1}. ${
              quiz.question
            }`}</div>
            <form>
              <FormControl component="fieldset" className={classes.formControl}>
                <RadioGroup
                  aria-label="quiz"
                  name="quiz"
                  value={selectedOption}
                  onChange={onOptionChange}
                >
                  {quiz.options.map((option, index) => {
                    return (
                      <FormControlLabel
                        className="h-50 option"
                        key={index}
                        value={option}
                        control={
                          <Radio
                            className="c-black"
                            checked={
                              isWrongAnswerMode
                                ? option === quiz.selectedOption
                                  ? true
                                  : false
                                : option === selectedOption
                                ? true
                                : false
                            }
                          />
                        }
                        label={option}
                        style={
                          !isWrongAnswerMode
                            ? selectedOption
                              ? selectedOption === option
                                ? option === quiz.correct_answer
                                  ? {
                                      backgroundColor: "#66DE93",
                                      pointerEvents: "none"
                                    }
                                  : {
                                      backgroundColor: "#FF616D",
                                      pointerEvents: "none"
                                    }
                                : { pointerEvents: "none" }
                              : null
                            : option === quiz.selectedOption
                            ? {
                                backgroundColor: "#FF616D",
                                pointerEvents: "none"
                              }
                            : option === quiz.correct_answer
                            ? {
                                backgroundColor: "#66DE93",
                                pointerEvents: "none"
                              }
                            : { pointerEvents: "none" }
                        }
                      />
                    );
                  })}
                </RadioGroup>
              </FormControl>
            </form>
          </div>
          {!isWrongAnswerMode ? (
            selectedOption && (
              <div className="f-d j-sb">
                <FormHelperText className="text-helper">
                  {quiz.isCorrect ? "정답입니다!" : "오답입니다!"}
                </FormHelperText>
                <div className="f-d">
                  {!quiz.isCorrect && !quiz.isOnceMore && (
                    <Button
                      className="normal-btn mr-20"
                      variant="contained"
                      onClick={onOnceMorePress}
                    >
                      {"한번 더!"}
                    </Button>
                  )}
                  <Button
                    className="normal-btn"
                    variant="contained"
                    onClick={isLast ? onFinishPress : onNextPress}
                  >
                    {isLast ? "종료" : "다음"}
                  </Button>
                </div>
              </div>
            )
          ) : (
            <div className="f-d j-sb">
              <TextField
                id="outlined-full-width"
                label="오답노트"
                className="text-field mr-20"
                placeholder="왜 틀렸는지 자신의 생각을 간단하게 메모해 보세요."
                fullWidth
                InputLabelProps={{
                  shrink: true
                }}
                variant="outlined"
                value={quiz.note ? quiz.note : ""}
                onChange={e => onWrongAnswerNoteTextChange(e, quiz)}
              />
              <div className="f-d">
                <Button
                  className="normal-btn"
                  variant="contained"
                  onClick={isLast ? onFinishPress : onNextPress}
                >
                  {isLast ? "저장 후 종료" : "다음"}
                </Button>
              </div>
            </div>
          )}
        </div>
      )}
      {isFinish && (
        <div
          className="result-form"
          data-testid={
            !isWrongAnswerMode ? "result-component" : "close-component"
          }
        >
          {!isWrongAnswerMode && (
            <React.Fragment>
              <div className="f-d j-sb">
                <div className="wd-30">
                  <div className="text-title mb-30">{"<결과>"}</div>
                  <div className="text-sub-title mb-10">{`소요 시간 : ${moment
                    .utc(time.duration)
                    .format("HH:mm:ss")}`}</div>
                  <div className="text-sub-title mb-10">{`정답 개수 : ${
                    quizzes.filter(quiz => quiz.isCorrect).length
                  }개`}</div>
                  <div className="text-sub-title">{`오답 개수 : ${
                    quizzes.filter(quiz => !quiz.isCorrect).length
                  }개`}</div>
                </div>
                <div className="wd-70">
                  <div className="text-title mb-30">{"<정 오답 비율>"}</div>
                  <div>
                    <Bar
                      data={data}
                      width={200}
                      height={250}
                      options={options}
                    />
                  </div>
                </div>
              </div>
              <div className="f-d j-fe a-c">
                <Button
                  className="normal-btn mr-20"
                  variant="contained"
                  onClick={onWrongAnswerNotePress}
                >
                  {"오답노트"}
                </Button>
                <Button
                  className="normal-btn"
                  variant="contained"
                  onClick={onRestartPress}
                >
                  {"다시풀기"}
                </Button>
              </div>
            </React.Fragment>
          )}
          {isWrongAnswerMode && (
            <div className="f-d j-c a-c h-100">
              <div className="text-title">{"수고하셨습니다."}</div>
            </div>
          )}
        </div>
      )}
    </div>
  );
};

export default Quiz;

const useStyles = makeStyles(theme => ({
  formControl: {
    margin: theme.spacing(3),
    marginLeft: 10,
    width: "100%"
  },
  button: {
    margin: theme.spacing(1, 1, 0, 0)
  }
}));
