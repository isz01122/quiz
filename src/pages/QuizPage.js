import React, { useEffect, useState } from "react";
import { Button, CircularProgress } from "@material-ui/core";
import decode from "html-entities-decode";
import moment from "moment";

import Quiz from "../components/Quiz";
import {
  correctAudio,
  incorrectAudio,
  startAudio,
  nextAudio,
  closeAudio
} from "../audios";
import API from "../services/API";

const QuizPage = () => {
  const [quizzes, setQuizzes] = useState([]);
  const [index, setIndex] = useState(0);
  const [isStart, setIsStart] = useState(false);
  const [isFinish, setIsFinish] = useState(false);
  const [isWrongAnswerMode, setIsWrongAnswerMode] = useState(false);
  const [inCorrectedQuizIds, setInCorrectedQuizIds] = useState([]);
  const [selectedOption, setSelectedOption] = useState("");
  const [time, setTime] = useState({
    startedAt: null,
    endedAt: null,
    duration: null
  });

  useEffect(() => {
    API.getQuizzes()
      .then(response => {
        const quizzes = brushUpQuizzes(response);
        setQuizzes(quizzes);
      })
      .catch(error => {
        console.log(error);
      });
  }, []);

  const brushUpQuizzes = response => {
    const quizzes = response.data.results.map((quiz, index) => {
      const _quiz = {
        ...quiz,
        id: index,
        question: decode(quiz.question),
        correct_answer: decode(quiz.correct_answer),
        incorrect_answers: quiz.incorrect_answers.map(option => decode(option)),
        options: [quiz.correct_answer, ...quiz.incorrect_answers]
          .sort(() => Math.random() - 0.5)
          .map(option => decode(option)),
        isCorrect: false,
        isOnceMore: false,
        note: null,
        selectedOption: null
      };
      return _quiz;
    });
    return quizzes;
  };

  const handleQuizStartPress = () => {
    new Audio(startAudio).play();
    setIsStart(true);
    setTime({
      ...time,
      startedAt: moment()
    });
  };

  const handleFinishPress = () => {
    const _endedAt = moment();
    if (isWrongAnswerMode) {
      new Audio(closeAudio).play();
    }
    setIndex(0);
    setSelectedOption("");
    setIsFinish(true);
    setTime({
      ...time,
      endedAt: _endedAt,
      duration: _endedAt.diff(time.startedAt)
    });
  };

  const handleOnceMorePress = () => {
    new Audio(nextAudio).play();
    const _quizzes = quizzes.map((quiz, idx) => {
      if (idx === index) {
        quiz.isOnceMore = true;
      }
      return quiz;
    });
    setSelectedOption("");
    setQuizzes(_quizzes);
  };

  const handleNextPress = () => {
    new Audio(nextAudio).play();
    setIndex(index + 1);
    setSelectedOption("");
  };

  const handleWrongAnswerNotePress = () => {
    new Audio(nextAudio).play();
    const _incorrectedIds = quizzes
      .filter(quiz => quiz.isCorrect === false)
      .map(quiz => quiz.id);
    if (_incorrectedIds.length > 0) {
      setInCorrectedQuizIds(_incorrectedIds);
      setIsWrongAnswerMode(true);
      setIsFinish(false);
    } else {
      alert("오답이 없습니다.");
    }
  };

  const handleWrongAnswerNoteTextChange = (e, _quiz) => {
    const _quizzes = quizzes.map(quiz => {
      if (quiz.id === _quiz.id) {
        quiz.note = e.target.value;
      }
      return quiz;
    });
    setQuizzes(_quizzes);
  };

  const handleRestartPress = () => {
    new Audio(nextAudio).play();
    let _quizzes = quizzes.map(quiz => {
      quiz.isCorrect = false;
      quiz.isOnceMore = false;
      return quiz;
    });
    setIndex(0);
    setSelectedOption("");
    setIsFinish(false);
    setTime({
      ...time,
      startedAt: moment(),
      endedAt: null,
      duration: null
    });
    setQuizzes(_quizzes);
  };

  const handleOptionChange = e => {
    const correctAnswer = quizzes[index].correct_answer;
    let seletedAnswer = e.target.value;
    let _quizzes = [];
    if (seletedAnswer === correctAnswer) {
      new Audio(correctAudio).play();
      _quizzes = quizzes.map((quiz, idx) => {
        if (idx === index) {
          quiz.isCorrect = true;
          quiz.selectedOption = seletedAnswer;
        }
        return quiz;
      });
    } else {
      new Audio(incorrectAudio).play();
      _quizzes = quizzes.map((quiz, idx) => {
        if (idx === index) {
          quiz.isCorrect = false;
          quiz.selectedOption = seletedAnswer;
        }
        return quiz;
      });
    }
    setQuizzes(_quizzes);
    setSelectedOption(seletedAnswer);
  };

  const buildQuizList = ids => {
    const _quizzes = quizzes.filter(quiz => ids.includes(quiz.id));
    return _quizzes;
  };

  return (
    <div className="main-container" data-testid={"quiz-page"}>
      {isStart && (
        <Quiz
          quizzes={
            isWrongAnswerMode ? buildQuizList(inCorrectedQuizIds) : quizzes
          }
          onFinishPress={handleFinishPress}
          onNextPress={handleNextPress}
          index={index}
          onOptionChange={e => handleOptionChange(e)}
          selectedOption={selectedOption}
          isFinish={isFinish}
          time={time}
          onRestartPress={handleRestartPress}
          onOnceMorePress={handleOnceMorePress}
          isWrongAnswerMode={isWrongAnswerMode}
          onWrongAnswerNotePress={handleWrongAnswerNotePress}
          onWrongAnswerNoteTextChange={(e, quiz) =>
            handleWrongAnswerNoteTextChange(e, quiz)
          }
        />
      )}
      {!isStart && (
        <Button
          className="quiz-start"
          variant="contained"
          onClick={handleQuizStartPress}
          disabled={quizzes.length === 0}
          data-testid={"quiz-start-button"}
        >
          {quizzes.length === 0 ? (
            <CircularProgress className="loading-spinner" size={50} />
          ) : (
            "퀴즈 풀기"
          )}
        </Button>
      )}
    </div>
  );
};

export default QuizPage;
