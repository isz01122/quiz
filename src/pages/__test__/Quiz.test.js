import { render, screen, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import QuizPage from "../QuizPage";
import Quiz from "../../components/Quiz";

afterEach(() => {
  cleanup();
});

test("퀴즈 페이지가 생성되어야함", () => {
  render(<QuizPage />);
  const quizPageElement = screen.getByTestId("quiz-page");
  expect(quizPageElement).toBeInTheDocument();
});

test("퀴즈 컴포넌트가 생성되어야함", () => {
  render(
    <Quiz
      quizzes={[
        {
          category: "My Category",
          type: "multiple",
          difficulty: "easy",
          question: "우리나라의 이름은?",
          correct_answer: "대한민국",
          incorrect_answers: ["미국", "일본", "중국"],
          id: 0,
          options: ["미국", "대한민국", "일본", "중국"],
          isCorrect: false,
          isOnceMore: false,
          note: null,
          selectedOption: null
        }
      ]}
      index={0}
      selectedOption={""}
      isFinish={false}
      time={{
        startedAt: null,
        endedAt: null,
        duration: null
      }}
      isWrongAnswerMode={false}
    />
  );
  const quizPageElement = screen.getByTestId("quiz-component");
  expect(quizPageElement).toBeInTheDocument();
});

test("퀴즈 컴포넌트 오답노트가 생성되어야함", () => {
  render(
    <Quiz
      quizzes={[
        {
          category: "My Category",
          type: "multiple",
          difficulty: "easy",
          question: "우리나라의 이름은?",
          correct_answer: "대한민국",
          incorrect_answers: ["미국", "일본", "중국"],
          id: 0,
          options: ["미국", "대한민국", "일본", "중국"],
          isCorrect: false,
          isOnceMore: false,
          note: null,
          selectedOption: null
        }
      ]}
      index={0}
      selectedOption={""}
      isFinish={false}
      time={{
        startedAt: null,
        endedAt: null,
        duration: null
      }}
      isWrongAnswerMode={true}
    />
  );
  const quizPageElement = screen.getByTestId(
    "quiz-component-wrong-answer-mode"
  );
  expect(quizPageElement).toBeInTheDocument();
});

test("퀴즈 결과 컴포넌트가 생성되어야함", () => {
  render(
    <Quiz
      quizzes={[
        {
          category: "My Category",
          type: "multiple",
          difficulty: "easy",
          question: "우리나라의 이름은?",
          correct_answer: "대한민국",
          incorrect_answers: ["미국", "일본", "중국"],
          id: 0,
          options: ["미국", "대한민국", "일본", "중국"],
          isCorrect: false,
          isOnceMore: false,
          note: null,
          selectedOption: null
        }
      ]}
      index={0}
      selectedOption={""}
      isFinish={true}
      time={{
        startedAt: null,
        endedAt: null,
        duration: null
      }}
      isWrongAnswerMode={false}
    />
  );
  const quizPageElement = screen.getByTestId("result-component");
  expect(quizPageElement).toBeInTheDocument();
});

test("퀴즈 종료화면 컴포넌트가 생성되어야함", () => {
  render(
    <Quiz
      quizzes={[
        {
          category: "My Category",
          type: "multiple",
          difficulty: "easy",
          question: "우리나라의 이름은?",
          correct_answer: "대한민국",
          incorrect_answers: ["미국", "일본", "중국"],
          id: 0,
          options: ["미국", "대한민국", "일본", "중국"],
          isCorrect: false,
          isOnceMore: false,
          note: null,
          selectedOption: null
        }
      ]}
      index={0}
      selectedOption={""}
      isFinish={true}
      time={{
        startedAt: null,
        endedAt: null,
        duration: null
      }}
      isWrongAnswerMode={true}
    />
  );
  const quizPageElement = screen.getByTestId("close-component");
  expect(quizPageElement).toBeInTheDocument();
});

test("matches snapshot", () => {
  const tree = renderer.create(<QuizPage />).toJSON();
  console.log(tree);
  expect(tree).toMatchSnapshot();
});
