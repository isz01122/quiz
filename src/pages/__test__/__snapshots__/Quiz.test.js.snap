// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="main-container"
  data-testid="quiz-page"
>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained quiz-start Mui-disabled Mui-disabled"
    data-testid="quiz-start-button"
    disabled={true}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={-1}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      <div
        className="MuiCircularProgress-root loading-spinner MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        role="progressbar"
        style={
          Object {
            "height": 50,
            "width": 50,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </div>
    </span>
  </button>
</div>
`;
